sdks/java/io/hadoop-file-system/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystem.java

```diff
@@ -189,7 +189,14 @@ public int read(ByteBuffer dst) throws IOException {
       if (closed) {
         throw new IOException("Channel is closed");
       }
-      return inputStream.read(dst);
+      // We avoid using the ByteBuffer based read for Hadoop because some FSInputStream
+      // implementations are not ByteBufferReadable,
+      // See https://issues.apache.org/jira/browse/HADOOP-14603
+      int read = inputStream.read(dst.array());
```

Comment:
I'd recommend looking at `PositionedReadable.readFully()`, which will handle partial reads of the dest buffer by reading in more data, etc. On a simple() read it could return any value, including 0.